### PR TITLE
New Clustering in Machine Learning Tutorial  (V 1.0)

### PR DESCRIPTION
**Pull request by khanteymoori**
Hi,
V 1.0 of clustering tutorial, includes hierarchical clustering, k-means, and DBSCAN.
Methods are applied to the Iris dataset, circles and moon samples.

@anuprulez, please have a look at it.

### CODE REVIEW

File: topics/statistics/tutorials/regression_machinelearning/tutorial.md [hunk 1]
```diff
@@ -56,6 +56,7 @@ Linear regression is a technique used to analyze a linear relationship between i
 
 In linear regression, our goal is to find the line that best models the path of the data points. Figure 3 shows the data set we used in Figure 2 with a line of best fit through it. The position of the line is determined by certain fitting coefficients (in this simple case, the gradient and intercept) and linear regression helps us pick appropriate values for these coefficients. In this example we have only one input variable and the problem is therefore simple linear regression. Note that in real problems we have more than one input variable. In this case, we call it multiple linear regression. Adding extra input variables just means that we’ll need to find more weights.
 
+
```
Comment (anuprulez) [inline topics/statistics/tutorials/regression_machinelearning/tutorial.md]: Why is there an empty line added?

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+hierarcial , k-means and DBSCAN algorithms. By following these steps, we learned how to perform clustering algorithms using clustering and plotting tools in Galaxy. Moreover, we visualized the results using multiple plots to analyse the clustering tasks. There are many other clustering approaches which can be tried out on these datasets to find how they perform. Different datasets can also be analysed using these algorithms. The clustering algorithms have some parameters which can be altered while performing the analyses to see if they affect the clustering or not. Although clustering is easy to implement, you need to take care of some important aspects like treating outliers in your data and making sure each cluster has sufficient population. Some data pre-processors can also be used to clean the datasets.
+
+
+
+
+
+
+
+
+
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: So many empty lines?

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+>    ![data](images/circles_clustering.png "Plot of Clustering Algorithms on Circles Data")
+
+>    ![data](images/moon_clustering.png "Plot of Clustering Algorithms on Moon Data")
+
+
+# Conclusion
+
+In this tutorial, we discussed the unsupervised learning task: clustering. Clustering is used to discover structure in unlabeled data. You learned about the
+hierarcial , k-means and DBSCAN algorithms. By following these steps, we learned how to perform clustering algorithms using clustering and plotting tools in Galaxy. Moreover, we visualized the results using multiple plots to analyse the clustering tasks. There are many other clustering approaches which can be tried out on these datasets to find how they perform. Different datasets can also be analysed using these algorithms. The clustering algorithms have some parameters which can be altered while performing the analyses to see if they affect the clustering or not. Although clustering is easy to implement, you need to take care of some important aspects like treating outliers in your data and making sure each cluster has sufficient population. Some data pre-processors can also be used to clean the datasets.
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: ` we learned how to perform clustering and visualize results using clustering and plotting tools respectively in Galaxy`

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: Can't say `the clustering is easy to implement`? Based on which evidence? Maybe some datasets are easy to cluster but not all

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: Please rephrase conclusion.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+title: 'Clustering in Machine Learning'
+zenodo_link: https://zenodo.org/...
+questions:
+- How to use clustering methods to categorized data in different clusters
+objectives:
+- Learn clustering background
+- Learn hierarchical clustering algorithm
+- Learn k-means clustering algorithm
+- Learn DBSCAN clustering algorithm
+- Apply clustering algorithms to different data sets
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: Clustering methods or clustering algorithms? Please maintain consistency

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+time_estimation: 2H
+contributors:
+- khanteymoori
+- anuprulez
+---
+
+# Introduction
+{:.no_toc}
+
+In this tutorial we will discuss an unsupervised learning task called clustering. The goal of unsupervised learning is to discover hidden structure or patterns in unlabeled training data.  Clustering is used to find groups of similar observations within a set of unlabeled data, such that members of the same group, or cluster, are more similar to each other by a given metric than they are to the members of the other clusters. Clustering is commonly used to explore a dataset. For example, in biology, clustering is used to find groups of genes with similar expression patterns. We cluster objects in multiple groups , such that objects within the same cluster are as similar as possible (i.e., high intra-class similarity), whereas objects from dierent clusters are as dissimilar as possible (i.e., low inter-class similarity).
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: `Clustering is commonly used to explore a dataset`. How?

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: First of all, what is a cluster and why is it important to find clusters in data?

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: `objects in multiple groups , such`. Extra space before a comma.

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: `dierent`. A typo

https://app.grammarly.com/ has a Chrome extension to check typos and grammar. Its a really useful tool :)

**Comment by anuprulez** [comment]
distances_clustering

Image is not displayed

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+In the above example, the best choice of number of clusters will be 4 as the red horizontal line in the dendrogram below covers maximum vertical distance AB.
+>    ![data](images/Hierarchical_clustering_2.png "Hierarchical Clustering")
+
+
+This algorithm has been implemented above using bottom up approach. It is also possible to follow top-down approach starting with all data points assigned in the same cluster and recursively performing splits till each data point is assigned a separate cluster.
+The decision of merging two clusters is taken on the basis of closeness of these clusters. 
+
+
+Now we will apply hierarchical clustering to iris data, and learn to measure its performance. At the first step, we should upload the Iris data set.
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: iris data, `Iris data set`?

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+>
+> 3. Rename to `Hierarchical scatter plot`
+
+
+>    ![data](images/hierarchical_scatter.png "Hierarchical Clustering Scatter Plot")
+
+
+# K-means Algorithm
+
+K-means clustering is the most commonly used unsupervised machine learning algorithm for partitioning a given data set into a set of k clusters, where k represents the number of groups pre-specified by the analyst.
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: who is the analyst here?

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+K-means clustering is the most commonly used unsupervised machine learning algorithm for partitioning a given data set into a set of k clusters, where k represents the number of groups pre-specified by the analyst. 
+In k-means clustering, each cluster is represented by its center or centroid which corresponds to the mean of points assigned to the cluster.
+The basic idea behind k-means clustering consists of defining clusters so that the total intra-cluster variation is minimized.
+The K-Means is popular because of its speed and scalability. There are several k-means algorithms available. The standard algorithm defines the total within-cluster variation as the sum of squared Euclidean distances between items and the corresponding centroid. K is a the hyperparameter of the algorithm and k-means algorithm can be summarized as follow:
+
+1. Specify the number of clusters (K) to be created (by the analyst)
+
+2. Select randomly k objects from the data set as the initial cluster centers or means
+
+3. Assigns each observation to their closest centroid, based on the Euclidean distance between the object and the centroid
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: Assign

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+
+4. For each of the k clusters update the cluster centroid by calculating the new mean values of all the data points in the cluster.
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: Update cluster centroid ...

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+> >
+> {: .solution}
+{: .question}
+
+
+# Clustering with the DBSCAN algorithm
+
+DBSCAN (Density-Based Spatial Clustering of Applications with Noise) is a popular clustering method and views clusters as areas of high density separated by areas of low density. Due to this rather generic view, clusters found by DBSCAN can be any shape, as opposed to k-means which assumes that clusters are convex shaped. The central component to the DBSCAN is the concept of core samples, which are samples that are in areas of high density. A cluster is therefore a set of core samples, each close to each other (measured by some distance measure) and a set of non-core samples that are close to a core sample (but are not themselves core samples). There are two parameters to the algorithm, Maximum neighborhood distance and Minimal core point density, which define formally what we mean when we say dense. Higher min_samples or lower eps indicate higher density necessary to form a cluster.
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: maximum

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: minimal

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: `min_samples`?

Comment (anuprulez) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: eps is not explained?

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+>        - *"Data point options"*: `User defined point options`
+>            - *"relative size of points"*: `2.0`
+>        - *"Plotting multiple groups"*: `Plot multiple groups of data on one plot`
+>            - *"column differentiating the different groups"*: `3`
+>            - *"Color schemes to differentiate your groups"*: `Set 2 - predefined color pallete`
+>
+> 2. **View** {% icon galaxy-eye%} the resulting plot:
+
+
+In the next steps, you can apply these three algorithms (hierarchical, k-means and DBSCAN) in simlar steps to moon and circles datasets. For the DBSCAN algorithm, the parameters are not the default parameters and you should set them as follows: for the circles data set ( Maximum neighborhood distance=0.2 and Minimal core point density=5) and for the moon datasets (Maximum neighborhood distance=0.3 and Minimal core point density=4). You can see the scatter plots of the clustering results in Figure 13 and Figure 14.
```
Comment (anuprulez) [inline topics/statistics/tutorials/clustering-machinelearning/tutorial.md]: Extra space: `data set ( M`

**Comment by anuprulez** [comment]
@khanteymoori I have added some comments. Please look at them. Thanks for working on it

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md (hunk 3 above)
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:473]: The comments are applied.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md (hunk 4 above)
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:13]: Done! all are changed to clustering algorithms

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md (hunk 5 above)
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:27]: The introduction is modified and the comments are applied.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,482 @@
...
+# Introduction
+{:.no_toc}
+
+In this tutorial we will discuss an unsupervised learning task called clustering. The goal of unsupervised learning is to discover hidden structure or patterns in unlabeled training data.  Clustering is used to find groups of similar observations within a set of unlabeled data, such that members of the same group, or cluster, are more similar to each other by a given metric than they are to the members of the other clusters. Clustering is commonly used to explore a dataset. For example, in biology, clustering is used to find groups of genes with similar expression patterns. We cluster objects in multiple groups , such that objects within the same cluster are as similar as possible (i.e., high intra-class similarity), whereas objects from dierent clusters are as dissimilar as possible (i.e., low inter-class similarity).
+
+We represent an observation as an n-dimensional vector. For example, assume that your training data consists of the samples plotted in the following Figure 1:
+
+>    ![data](images/data_before_clustering.png "Sample data before clustering")
+
+Clustering might reveal the following four groups, indicated by different colors:
```
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:33]: done!

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+
+>    ![data](images/data_after_clustering.png "Sample data after clustering")
+
+
+Broadly speaking, clustering can be divided into two subgroups :
+
+- Hard Clustering: In hard clustering, each data point either belongs to a cluster completely or not.
```
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:40]: is removed

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+	
+- Soft Clustering: In soft clustering, instead of putting each data point into a separate cluster, a probability or likelihood of that data point to be in those clusters is assigned. 
+
+The goal of clustering is to determine the internal grouping in a set of unlabeled data. But how to decide what constitutes a good clustering? It can be shown that there is no absolute best criterion which would be independent of the final aim of the clustering. Consequently, it is the user who should supply this criterion, in such a way that the result of the clustering will suit their needs.
```
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:44]: The problem with the image is done. 
The sentence is modified.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+>
+> 1. TOC
+> {:toc}
+>
+{: .agenda}
+
+
+# Types of clustering algorithms
+
+Since clustering is a subjective task, there are many algorithms for data clustering. Every method follows a different set of rules for defining the similarity among the There are  many clustering algorithms known. But few of the algorithms are used popularly and we can categorized them as follows:
```
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:60]: problem is solved

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 16]
```diff
@@ -0,0 +1,482 @@
...
+
+ - Centroid models: These are iterative clustering algorithms in which the notion of similarity is derived by the closeness of a data point to the centroid of the clusters. K-means clustering algorithm is a popular algorithm that falls into this category. In these models, the number of clusters required at the end have to be mentioned beforehand, which makes it important to have prior knowledge of the dataset. These models run iteratively to find the local optima.
+
+ - Density Models: These models search the data space for areas of varied density of data points in the data space. It isolates various different density regions and assign the data points within these regions in the same cluster. Popular example of density models is DBSCAN.
+
+ - Distribution models: These clustering models are based on the notion of how probable is it that all data points in the cluster belong to the same distribution, for example normal, Gaussian. These models often suffer from overfitting. A popular example of these models is Expectation-maximization algorithm which uses multivariate normal distributions.
```
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:68]: modified!

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md [hunk 17]
```diff
@@ -0,0 +1,482 @@
...
+ > >
+ > > 1. The distance between a point to itself is zero.
+ > > 1. The distance between point *a* to point *b* is the same as the distance between point *b* to point *a* using the Euclidean distance metric.
+ > >
+ > {: .solution }
+ >
+ {: .question }
+
+Other dissimilarity measures exist such as correlation-based distances, which is widely used for gene expression data analyses. Correlation-based distance considers two objects to be similar if their features are highly correlated, even though the observed values may be far apart in terms of Euclidean distance. The distance between two objects is 0 when they are perfectly correlated. Pearson’s correlation is quite sensitive to outliers. This does not matter
+when clustering samples, because the correlation is over thousands of genes. During clustering genes, it is important to be aware of the possible impact of outliers. This
```
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:104]: outliers are recognized as the points belonging to none of the clusters. clustering without removing the outliers leads to different clusters.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md (hunk 10 above)
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:347]: eps and min_samples are explained.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md (hunk 7 above)
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:253]: I mean the user.

File: topics/statistics/tutorials/clustering-machinelearning/tutorial.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (khanteymoori) [reply topics/statistics/tutorials/clustering-machinelearning/tutorial.md:27]: more explanation is added

**Comment by khanteymoori** [comment]
@anuprulez 
Thanks for comments. I have applied them to the new version of the tutorial.
Unfortunately, I had a problem with this branch in local repository. 
So I will send a new PR with these changes.

**Comment by nsoranzo** [comment]
Closing in favour of #1870.

**Pull request closed without merging**